In [1]:
import pandas as pd
import statsmodels.api as sm
import numpy as np
import matplotlib.pyplot as plt
from linearmodels.iv import IV2SLS

In [2]:
# Exercise 1
wage2 = pd.read_stata("./stata/WAGE2.DTA")
X = sm.add_constant(wage2[["sibs"]])
model = sm.OLS(wage2.lwage, X, missing="drop").fit()
model.summary()

/home/systemchalk/anaconda3/lib/python3.9/site-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  lwage   R-squared:                       0.023
Model:                            OLS   Adj. R-squared:                  0.022
Method:                 Least Squares   F-statistic:                     22.31
Date:                Wed, 23 Mar 2022   Prob (F-statistic):           2.68e-06
Time:                        02:20:12   Log-Likelihood:                -506.59
No. Observations:                 935   AIC:                             1017.
Df Residuals:                     933   BIC:                             1027.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          6.8611      0.022    310.771      0.000       6.818       6.904
sibs          -0.0279      0.006     -4.723      0.000      -0.039      -0.016
==============================================================================
Omnibus:                       22.562   Durbin-Watson:                   1.726
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               26.529
Skew:                          -0.306   Prob(JB):                     1.74e-06
Kurtosis:                       3.553   Cond. No.                         6.33
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [3]:
X = sm.add_constant(wage2[["brthord"]])
model = sm.OLS(wage2.educ, X, missing="drop").fit()
model.summary()

/home/systemchalk/anaconda3/lib/python3.9/site-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   educ   R-squared:                       0.042
Model:                            OLS   Adj. R-squared:                  0.041
Method:                 Least Squares   F-statistic:                     37.29
Date:                Wed, 23 Mar 2022   Prob (F-statistic):           1.55e-09
Time:                        02:20:13   Log-Likelihood:                -1861.9
No. Observations:                 852   AIC:                             3728.
Df Residuals:                     850   BIC:                             3737.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         14.1494      0.129    109.962      0.000      13.897      14.402
brthord       -0.2826      0.046     -6.106      0.000      -0.373      -0.192
==============================================================================
Omnibus:                       77.090   Durbin-Watson:                   1.784
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               53.992
Skew:                           0.506   Prob(JB):                     1.89e-12
Kurtosis:                       2.294   Cond. No.                         5.28
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [4]:
IV2SLS(wage2.lwage, np.ones((wage2.shape[0], 1)), wage2.educ, wage2.brthord).fit(cov_type="unadjusted")

/home/systemchalk/anaconda3/lib/python3.9/site-packages/linearmodels/shared/exceptions.py:35: MissingValueWarning: 
Inputs contain missing values. Dropping rows with missing observations.
  warnings.warn(missing_value_warning_msg, MissingValueWarning)


IV-2SLS Estimation Summary                          
==============================================================================
Dep. Variable:                  lwage   R-squared:                     -0.0286
Estimator:                    IV-2SLS   Adj. R-squared:                -0.0298
No. Observations:                 852   F-statistic:                    16.667
Date:                Wed, Mar 23 2022   P-value (F-stat)                0.0000
Time:                        02:20:13   Distribution:                  chi2(1)
Cov. Estimator:            unadjusted                                         
                                                                              
                             Parameter Estimates                              
==============================================================================
            Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
------------------------------------------------------------------------------
exog           5.0304     0.4324     11.633     0.0000      4.1828      5.8780
educ           0.1306     0.0320     4.0825     0.0000      0.0679      0.1934
==============================================================================

Endogenous: educ
Instruments: brthord
Unadjusted Covariance (Homoskedastic)
Debiased: False
IVResults, id: 0x7fe3fc116f40

In [5]:
X = sm.add_constant(wage2[["sibs", "brthord"]])
model = sm.OLS(wage2.educ, X, missing="drop").fit()
# We need the fitted values for part vi (first stage results)
educ_hat = model.fittedvalues
model.summary()

/home/systemchalk/anaconda3/lib/python3.9/site-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   educ   R-squared:                       0.058
Model:                            OLS   Adj. R-squared:                  0.056
Method:                 Least Squares   F-statistic:                     26.29
Date:                Wed, 23 Mar 2022   Prob (F-statistic):           8.33e-12
Time:                        02:20:13   Log-Likelihood:                -1854.6
No. Observations:                 852   AIC:                             3715.
Df Residuals:                     849   BIC:                             3729.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         14.2965      0.133    107.260      0.000      14.035      14.558
sibs          -0.1529      0.040     -3.834      0.000      -0.231      -0.075
brthord       -0.1527      0.057     -2.675      0.008      -0.265      -0.041
==============================================================================
Omnibus:                       69.406   Durbin-Watson:                   1.795
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               51.104
Skew:                           0.497   Prob(JB):                     8.00e-12
Kurtosis:                       2.329   Cond. No.                         8.46
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [6]:
X = sm.add_constant(wage2[["sibs"]])
IV2SLS(wage2.lwage, X, wage2.educ, wage2.brthord).fit(cov_type="unadjusted")

/home/systemchalk/anaconda3/lib/python3.9/site-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)
/home/systemchalk/anaconda3/lib/python3.9/site-packages/linearmodels/shared/exceptions.py:35: MissingValueWarning: 
Inputs contain missing values. Dropping rows with missing observations.
  warnings.warn(missing_value_warning_msg, MissingValueWarning)


IV-2SLS Estimation Summary                          
==============================================================================
Dep. Variable:                  lwage   R-squared:                     -0.0543
Estimator:                    IV-2SLS   Adj. R-squared:                -0.0568
No. Observations:                 852   F-statistic:                    21.871
Date:                Wed, Mar 23 2022   P-value (F-stat)                0.0000
Time:                        02:20:13   Distribution:                  chi2(2)
Cov. Estimator:            unadjusted                                         
                                                                              
                             Parameter Estimates                              
==============================================================================
            Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
------------------------------------------------------------------------------
const          4.9385     1.0538     4.6863     0.0000      2.8731      7.0040
sibs           0.0021     0.0173     0.1217     0.9031     -0.0319      0.0361
educ           0.1370     0.0745     1.8376     0.0661     -0.0091      0.2831
==============================================================================

Endogenous: educ
Instruments: brthord
Unadjusted Covariance (Homoskedastic)
Debiased: False
IVResults, id: 0x7fe3fc151b20

In [7]:
educ_hat.corr(wage2.sibs)

-0.929481763991843

C1.i When plugging in sibs in as a proxy for education, the coefficent is negative, sufficient to show it is not the same as the IV (0.122, from the text). Specifically this regression says that, without controlling for other factors, an additional sibling is associated with a little under 3% less in wages (though we shouldn't place too much emphasis on this as sibs is almost certianly correlated with other factors)

C1.ii One intuitive reason why education might be negatively associated with birth order is simply the matter of resource constraints (older children may have money set aside for education while younger children may not). The regression shows a statistically significant negative relationship between education and birth order, with each new position in the rank corresponding to a little over a quarter of a year less in education.

C1.iii Results reported above. When using birth order as an instrument for education, we see a larger coefficient for education.

C1.iv We are assuming that $\pi_2 \neq 0$ (that is, brthord isn't zero). The regression allows us to reject the null hypothesis that this value is zero, and so the identification assumption holds.

C1.v Results above. The standard error is considerably higher than the one estimated previously (0.0745 vs. 0.0320), enough to make it lose significance. Sibs has a small standard error, but also a small coefficient, meaning we have no evidence that it is significantly different from zero.

C1.vi There is a strong negative correlation between the fitted values of educ and sibs (-0.93). Recalling from section 3 that multicollinearity is a problem for 2SLS, we have an explanation for our large standard errors.

In [8]:
# Exercise 2
fertil2 = pd.read_stata("./stata/FERTIL2.DTA")
X = sm.add_constant(fertil2[["educ", "age", "agesq"]])
model = sm.OLS(fertil2.children, X, missing="drop").fit()
model.summary()

/home/systemchalk/anaconda3/lib/python3.9/site-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               children   R-squared:                       0.569
Model:                            OLS   Adj. R-squared:                  0.568
Method:                 Least Squares   F-statistic:                     1915.
Date:                Wed, 23 Mar 2022   Prob (F-statistic):               0.00
Time:                        12:01:58   Log-Likelihood:                -7835.6
No. Observations:                4361   AIC:                         1.568e+04
Df Residuals:                    4357   BIC:                         1.570e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -4.1383      0.241    -17.200      0.000      -4.610      -3.667
educ          -0.0906      0.006    -15.298      0.000      -0.102      -0.079
age            0.3324      0.017     20.088      0.000       0.300       0.365
agesq         -0.0026      0.000     -9.651      0.000      -0.003      -0.002
==============================================================================
Omnibus:                      203.406   Durbin-Watson:                   1.868
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              715.951
Skew:                           0.017   Prob(JB):                    3.41e-156
Kurtosis:                       4.985   Cond. No.                     1.07e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.07e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [9]:
X = sm.add_constant(fertil2[["frsthalf", "age", "agesq"]])
model = sm.OLS(fertil2.educ, X, missing="drop").fit()
model.summary()

/home/systemchalk/anaconda3/lib/python3.9/site-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   educ   R-squared:                       0.108
Model:                            OLS   Adj. R-squared:                  0.107
Method:                 Least Squares   F-statistic:                     175.2
Date:                Wed, 23 Mar 2022   Prob (F-statistic):          3.01e-107
Time:                        12:09:38   Log-Likelihood:                -11905.
No. Observations:                4361   AIC:                         2.382e+04
Df Residuals:                    4357   BIC:                         2.384e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          9.6929      0.598     16.207      0.000       8.520      10.865
frsthalf      -0.8523      0.113     -7.554      0.000      -1.073      -0.631
age           -0.1080      0.042     -2.568      0.010      -0.190      -0.026
agesq         -0.0005      0.001     -0.730      0.466      -0.002       0.001
==============================================================================
Omnibus:                       51.991   Durbin-Watson:                   1.352
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               56.555
Skew:                           0.233   Prob(JB):                     5.24e-13
Kurtosis:                       3.306   Cond. No.                     1.05e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.05e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [10]:
X = sm.add_constant(fertil2[["age", "agesq"]])
IV2SLS(fertil2.children, X, fertil2.educ, fertil2.frsthalf).fit(cov_type="unadjusted")

/home/systemchalk/anaconda3/lib/python3.9/site-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


IV-2SLS Estimation Summary                          
==============================================================================
Dep. Variable:               children   R-squared:                      0.5502
Estimator:                    IV-2SLS   Adj. R-squared:                 0.5499
No. Observations:                4361   F-statistic:                    5300.2
Date:                Wed, Mar 23 2022   P-value (F-stat)                0.0000
Time:                        12:14:21   Distribution:                  chi2(3)
Cov. Estimator:            unadjusted                                         
                                                                              
                             Parameter Estimates                              
==============================================================================
            Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
------------------------------------------------------------------------------
const         -3.3878     0.5479    -6.1833     0.0000     -4.4617     -2.3139
age            0.3236     0.0179     18.128     0.0000      0.2886      0.3586
agesq         -0.0027     0.0003    -9.5589     0.0000     -0.0032     -0.0021
educ          -0.1715     0.0532    -3.2264     0.0013     -0.2757     -0.0673
==============================================================================

Endogenous: educ
Instruments: frsthalf
Unadjusted Covariance (Homoskedastic)
Debiased: False
IVResults, id: 0x7fe3f90c7ac0

In [11]:
X = sm.add_constant(fertil2[["educ", "age", "agesq", "electric", "tv", "bicycle"]])
model = sm.OLS(fertil2.children, X, missing="drop").fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               children   R-squared:                       0.576
Model:                            OLS   Adj. R-squared:                  0.575
Method:                 Least Squares   F-statistic:                     984.9
Date:                Wed, 23 Mar 2022   Prob (F-statistic):               0.00
Time:                        12:15:22   Log-Likelihood:                -7789.3
No. Observations:                4356   AIC:                         1.559e+04
Df Residuals:                    4349   BIC:                         1.564e+04
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -4.3898      0.240    -18.267      0.000      -4.861      -3.919
educ          -0.0767      0.006    -12.075      0.000      -0.089      -0.064
age            0.3402      0.016     20.692      0.000       0.308       0.372
agesq         -0.0027      0.000    -10.010      0.000      -0.003      -0.002
electric      -0.3027      0.076     -3.974      0.000      -0.452      -0.153
tv            -0.2531      0.091     -2.768      0.006      -0.432      -0.074
bicycle        0.3179      0.049      6.440      0.000       0.221       0.415
==============================================================================
Omnibus:                      196.639   Durbin-Watson:                   1.891
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              675.502
Skew:                           0.009   Prob(JB):                    2.07e-147
Kurtosis:                       4.929   Cond. No.                     1.08e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.08e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [12]:
X = sm.add_constant(fertil2[["age", "agesq", "electric", "tv", "bicycle"]])
IV2SLS(fertil2.children, X, fertil2.educ, fertil2.frsthalf).fit(cov_type="unadjusted")

/home/systemchalk/anaconda3/lib/python3.9/site-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)
/home/systemchalk/anaconda3/lib/python3.9/site-packages/linearmodels/shared/exceptions.py:35: MissingValueWarning: 
Inputs contain missing values. Dropping rows with missing observations.
  warnings.warn(missing_value_warning_msg, MissingValueWarning)


IV-2SLS Estimation Summary                          
==============================================================================
Dep. Variable:               children   R-squared:                      0.5577
Estimator:                    IV-2SLS   Adj. R-squared:                 0.5571
No. Observations:                4356   F-statistic:                    5539.2
Date:                Wed, Mar 23 2022   P-value (F-stat)                0.0000
Time:                        12:15:44   Distribution:                  chi2(6)
Cov. Estimator:            unadjusted                                         
                                                                              
                             Parameter Estimates                              
==============================================================================
            Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
------------------------------------------------------------------------------
const         -3.5913     0.6446    -5.5717     0.0000     -4.8547     -2.3280
age            0.3281     0.0190     17.231     0.0000      0.2908      0.3655
agesq         -0.0027     0.0003    -9.8509     0.0000     -0.0033     -0.0022
electric      -0.1065     0.1658    -0.6424     0.5206     -0.4316      0.2185
tv            -0.0026     0.2091    -0.0122     0.9902     -0.4123      0.4072
bicycle        0.3321     0.0515     6.4499     0.0000      0.2312      0.4330
educ          -0.1640     0.0655    -2.5045     0.0123     -0.2923     -0.0357
==============================================================================

Endogenous: educ
Instruments: frsthalf
Unadjusted Covariance (Homoskedastic)
Debiased: False
IVResults, id: 0x7fe3f91b0fa0

C2.i Results above. Holding age fixed, each year of education results in a little less than .1 fewer children. For a group of 100 women who get another year of education, we would expect them to have about 9 fewer children.

C2.ii Regressing on the reduce form produces an estimate that allows us to reject the null hypothesis that the coefficient for firsthalf is zero. Since we have assumed it is uncorrelated with the error, this shows us we have a reasonable IV.

C2.iii Results above. The effect is larger than OLS while retaining significance (though it does have a wider confidence interval)

C3.iv Results above. The magnitude of education falls after these factors are accounted for (though it is still large). TV ownership is associated with fewer children (0.25 fewer for OLS, 0.0026 for 2SLS), likely because it is a proxy for higher incomes which are associated with fewer children.

In [ ]:
# Exercise 3